src/graph/graph.c

```diff
@@ -25,22 +25,30 @@ void _Graph_LeaveCriticalSection(Graph *g) {
     pthread_mutex_unlock(&g->_mutex);
 }
 
+/* Force execution of all pending operations on a matrix. */
+static inline void _Graph_SynchronizeMatrix(GrB_Matrix m) {
+    GrB_Index nvals;
+    GrB_Matrix_nvals(&nvals, m);
+}
+
 /* Resize given matrix, such that its number of row and columns
  * matches the number of nodes in the graph. */
 void _Graph_ResizeMatrix(const Graph *g, GrB_Matrix m) {
+    _Graph_EnterCriticalSection((Graph *)g);
     GrB_Index n_rows;
 
     GrB_Matrix_nrows(&n_rows, m);
     if(n_rows != Graph_NodeCount(g)) {
-        _Graph_EnterCriticalSection((Graph *)g);
-        {
-            // Double check now that we're in critical section.
-            GrB_Matrix_nrows(&n_rows, m);
-            if(n_rows != Graph_NodeCount(g))
-                assert(GxB_Matrix_resize(m, Graph_NodeCount(g), Graph_NodeCount(g)) == GrB_SUCCESS);
-        }
-        _Graph_LeaveCriticalSection((Graph *)g);
+        // Double check now that we're in critical section.
+        GrB_Matrix_nrows(&n_rows, m);
+        if(n_rows != Graph_NodeCount(g))
+            assert(GxB_Matrix_resize(m, Graph_NodeCount(g), Graph_NodeCount(g)) == GrB_SUCCESS);
     }
+
+    // Force updates to matrix
+    _Graph_SynchronizeMatrix(m);
+
+    _Graph_LeaveCriticalSection((Graph *)g);
```

Comment:
I do not love sandwiching this entire function within a mutex, but it guarantees that any retrieved matrices have no pending operations, so for the moment it is my solution over analyzing all matrix operations to see which are not thread-safe.

Let's discuss alternatives later!

Reply:
doesn't a matrix maintains an internal property which we can check to see if it has pending updates?
I think that wrapping this entire function with a critical section will impact performance as mutex acquisition is expensive operation